src/core/thread/mle_router.cpp

```diff
@@ -4482,6 +4483,66 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
+
+    OT_ASSERT(IsDisabled() || IsDetached() || IsRouter() || IsLeader());
```

Comment:
```suggestion
```
nit: looks unnecessary since we already exit for IsChild().

Reply:
I think the invariant "it must be and only be one of the 5 states" isn't very obvious, so I added this assertion. But if you think it's not necessary, I'm OK to remove it.